drivers/gpio/gpio_nrfx.c

```diff
@@ -0,0 +1,455 @@
+	return 0;
+}
+
+static int gpio_nrfx_manage_callback(struct device *port,
+				     struct gpio_callback *callback,
+				     bool set)
+{
+	_gpio_manage_callback(&get_port_data(port)->callbacks, callback, set);
+	return 0;
+}
+
+static int gpio_nrfx_enable_callback(struct device *port,
+				     int access_op,
+				     u32_t pin)
+{
+	struct gpio_nrfx_data *data = get_port_data(port);
+	u8_t pins_affected = 1;
+	int res = 0;
+
+	if (access_op == GPIO_ACCESS_BY_PORT) {
+		pins_affected = popcount(pin);
+	} else {
+		pin = BIT(pin);
+	}
+
+	for (size_t i = 0; i < pins_affected; i++) {
+		u32_t cur_pin = find_lsb_set(pin) - 1;
+
+		pin &= ~BIT(cur_pin);
+
+		if (!(data->pin_int_en & BIT(cur_pin))) {
+			return -EINVAL;
+		}
+
+		data->int_en |= BIT(cur_pin);
+
+		res = gpiote_pin_int_cfg(port, cur_pin);
+		if (res != 0) {
+			return res;
+		}
+	}
+
+	return res;
+}
+
+static int gpio_nrfx_disable_callback(struct device *port,
+				      int access_op,
+				      u32_t pin)
+{
+	struct gpio_nrfx_data *data = get_port_data(port);
+	u8_t pins_affected = 1;
+	int res = 0;
+
+	if (access_op == GPIO_ACCESS_BY_PORT) {
+		pins_affected = popcount(pin);
+	} else {
+		pin = BIT(pin);
+	}
+
+	for (size_t i = 0; i < pins_affected; i++) {
+		u32_t cur_pin = find_lsb_set(pin) - 1;
+
+		pin &= ~BIT(cur_pin);
+
+		data->int_en &= ~BIT(cur_pin);
+
+		res = gpiote_pin_int_cfg(port, cur_pin);
+		if (res != 0) {
+			return res;
+		}
+	}
+
+	return res;
+}
+
+static const struct gpio_driver_api gpio_nrfx_drv_api_funcs = {
+	.config = gpio_nrfx_config,
+	.write = gpio_nrfx_write,
+	.read = gpio_nrfx_read,
+	.manage_callback = gpio_nrfx_manage_callback,
+	.enable_callback = gpio_nrfx_enable_callback,
+	.disable_callback = gpio_nrfx_disable_callback
+};
+
+static u32_t get_trigger_pins(struct device *port)
+{
+	if (port == NULL) {
+		return 0;
+	}
+	struct gpio_nrfx_data *data = get_port_data(port);
+
+	u32_t port_in = nrf_gpio_port_in_read(get_port_cfg(port)->port);
+	u32_t xor = (port_in ^ data->active_level ^ data->polarity);
+	u32_t out = ~xor & data->int_en & data->pin_int_en
+		    & ~data->trig_edge & ~data->double_edge;
+
+	return out;
+}
+
+#ifdef CONFIG_GPIO_NRFX_P0
+DEVICE_DECLARE(gpio_nrfx_p0);
+#endif
+#ifdef CONFIG_GPIO_NRFX_P1
+DEVICE_DECLARE(gpio_nrfx_p1);
+#endif
+
+static struct device *port_list[GPIO_COUNT] = {
+
+#ifdef CONFIG_GPIO_NRFX_P0
+	DEVICE_GET(gpio_nrfx_p0),
+#else
+	NULL,
+#endif
+#ifdef CONFIG_GPIO_NRFX_P1
+	DEVICE_GET(gpio_nrfx_p1),
+#else
+	NULL,
+#endif
+};
+
+static void gpiote_event_handler(void)
+{
+	u32_t pins[GPIO_COUNT];
+	bool clear_port_evt = true;
+
+	memset(pins, 0, sizeof(pins));
+
+	if (nrf_gpiote_event_is_set(NRF_GPIOTE_EVENTS_PORT)) {
+		for (size_t i = 0; i < GPIO_COUNT; i++) {
+			pins[i] |= get_trigger_pins(port_list[i]);
+		}
+	}
+
+	for (size_t i = 0; i < GPIO_COUNT; i++) {
+		if (pins[i] != 0) {
+			clear_port_evt = false;
+			break;
+		}
+	}
+	if (clear_port_evt) {
+		nrf_gpiote_event_clear(NRF_GPIOTE_EVENTS_PORT);
```

Comment:
Sorry, I don't get it. Why not clear the PORT event? It will generate the IRQ again, won't it?